test/eigen.jl

```diff
@@ -82,6 +82,15 @@ using StaticArrays, Test, LinearAlgebra
             E = eigen(Symmetric(A, uplo))
             @test eigvecs(E) * SDiagonal(eigvals(E)) * eigvecs(E)' ≈ A
         end
+
+        m1_a = randn(2,2)
+        m1_a = m1_a*m1_a'
+        m1 = SMatrix{2,2}(m1_a)
+        m2_a = randn(2,2)
+        m2_a = m2_a*m2_a'
+        m2 = SMatrix{2,2}(m2_a)
+        @test eigvals(m1, m2) ≈ eigvals(m1_a, m2_a)
+        @test isa(eigvals(m1, m2), SVector{2})
```

Comment:
Makes sense but I think Jeff will not be pleased :slightly_smiling_face:  https://discourse.julialang.org/t/roadmap-for-a-faster-time-to-first-plot/22956/11

Reply:
Heh. I think that comment by Jeff more refers to people relying on `return_type` for functionality. Here we use it for testing only (inside `@inferred`), but never call it in the `src` directory, with one exception which is a minor crime also committed by Base.

Unfortunately it's exceedingly easy to break inference or inlining when refactoring this package, hence the unfortunately large amount of `@inferred`.

Feel free not to change these tests in any case.

Reply:
I'll try then to make `@inferred_union` that detects if the result is inferred as a union of concrete types (I guess that's the only case where `Union` can be significantly faster than `Any`?)

Reply:
Yeah definitely a union of concrete types.

I guess it might be rather simpler and more useful to make an `@inferred_type` macro which just calls `Base.return_types` (a much simplified version of https://github.com/JuliaLang/julia/blob/c7e9d9f8ade647582c6635c8c9a587f2360af404/stdlib/Test/src/Test.jl#L1384 for example) to return the actual inferred type. Then we can check more specifically that we get `Union{SVector{2,Float64}, SVector{2,ComplexF64}}` here.

Reply:
Julia since 1.2 has `@inferred` that can allow unions so we could just use that, right?

Reply:
I could just write `@inferred_maybe_allow A B` that expands to `@inferred B` on Julia 1.0 and 1.1 and to `@inferred A B` on Julia 1.2 and newer. Does that sound right?

Reply:
Oh I didn't know about that (despite linking to the source)! Let's definitely use that then.

The semantic is a bit confusing, I guess we name the type which doesn't appear in the output

```
@inferred SVector{2,ComplexF64} eigvals(m1, m2)
```

Reply:
Well it's technically not inferred so that may break on <1.2. Might be easiest to just add our own version of the two-argument `@inferred` macro method for Julia < 1.2, which just expands to `B` in 1.0 and 1.1. Unless Compat did it for us.

Reply:
Compat doesn't have this macro so I've just introduced `@inferred_maybe_allow` that expands differently on Julia <1.2 and >=1.2. Does that look OK?

Reply:
Looks great thanks. I'm tempted to rename it to `@inferred(a,b)` (macro overloading based on number of arguments works, so if the 2-arg form doesn't exist in earlier julia versions we can fairly cleanly supply it).

Reply:
OK, I'll rename it.

Reply:
Oh, the PR is merged, so maybe another time.